crates/vm/src/expr.rs

```diff
@@ -982,57 +975,61 @@ fn select_best_index<'a>(
+            // since those are typically not answered using an index.
+            for cmp in [OpCmp::Eq, OpCmp::Lt, OpCmp::LtEq, OpCmp::Gt, OpCmp::GtEq] {
+                // For a single column index,
+                // we want to avoid the `ProductValue` indirection of below.
+                for FieldValue { cmp, value, field, .. } in
+                    fields_map.remove(&(col_list.head(), cmp)).into_iter().flatten()
+                {
+                    found.push(make_index_arg(cmp, col_list, value.clone()));
+                    fields_indexed.insert((field, cmp));
+                }
             }
-        } else if col_list
-            .iter()
-            // (2) Ensure that every col has a field.
-            .all(|col| fields_map.get(&(col, cmp)).filter(|fs| !fs.is_empty()).is_some())
-        {
-            // We've ensured `col_list ⊆ columns_of(field_map(cmp))`.
-            // Construct the value to compare against.
-            let mut elems = Vec::with_capacity(col_list.len() as usize);
-            for col in col_list.iter() {
-                // Retrieve the field for this (col, cmp) key.
-                // Remove the map entry if the list is empty now.
-                let Entry::Occupied(mut entry) = fields_map.entry((col, cmp)) else {
-                    // We ensured in (2) that the map is occupied for `(col, cmp)`.
-                    unreachable!()
-                };
-                let fields = entry.get_mut();
-                // We ensured in (2) that `fields` is non-empty.
-                let field = fields.pop().unwrap();
-                if fields.is_empty() {
-                    // Remove the entry so that (1) works.
-                    entry.remove();
+        } else {
+            // We have a multi column index.
+            // Try to fit constraints `c_0 = v_0, ..., c_n = v_n` to this index.
+            //
+            // For the time being, we restrict multi-col index scans to `=` only.
+            // This is what our infrastructure is set-up to handle soundly.
+            // To extend this support to ranges requires deeper changes.
+            // TODO(Centril, 2024-05-30): extend this support to ranges.
+            let cmp = OpCmp::Eq;
+
+            // Compute the minimum number of `=` constraints that every column in the index has.
+            let mut min_all_cols_num_eq = col_list
+                .iter()
+                .map(|col| fields_map.get(&(col, cmp)).map_or(0, |fs| fs.len()))
+                .min()
+                .unwrap_or_default();
+
+            // For all of these sets of constraints,
+            // construct the value to compare against.
+            while min_all_cols_num_eq > 0 {
+                let mut elems = Vec::with_capacity(col_list.len() as usize);
+                for col in col_list.iter() {
+                    // Cannot panic as `min_all_cols_num_eq > 0`.
+                    let field = pop_multimap(&mut fields_map, (col, cmp)).unwrap();
```

Comment:
Why not use `let Some(...)` here to skip the two passes for the search of `col, cmp`?